staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go

```diff
@@ -48,48 +54,38 @@ type timeBudgetImpl struct {
 	sync.Mutex
 	budget time.Duration
 
-	refresh   time.Duration
 	maxBudget time.Duration
+	clock     clock.Clock
+	limiter   *rate.Limiter
 }
 
-func newTimeBudget(stopCh <-chan struct{}) timeBudget {
+func newTimeBudget() timeBudget {
+	clock := clock.RealClock{}
+	limiter := rate.NewLimiter(rate.Limit(1), burstSize)
 	result := &timeBudgetImpl{
+		clock:     clock,
 		budget:    time.Duration(0),
-		refresh:   refreshPerSecond,
 		maxBudget: maxBudget,
+		limiter:   limiter,
 	}
-	go result.periodicallyRefresh(stopCh)
+	clock.Sleep(limiter.ReserveN(clock.Now(), burstSize).DelayFrom(clock.Now()))
 	return result
 }
 
-func (t *timeBudgetImpl) periodicallyRefresh(stopCh <-chan struct{}) {
-	ticker := time.NewTicker(time.Second)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-ticker.C:
-			t.Lock()
-			if t.budget = t.budget + t.refresh; t.budget > t.maxBudget {
-				t.budget = t.maxBudget
-			}
-			t.Unlock()
-		case <-stopCh:
-			return
-		}
-	}
-}
-
 func (t *timeBudgetImpl) takeAvailable() time.Duration {
-	t.Lock()
-	defer t.Unlock()
-	result := t.budget
+	duration := 0
+	for t.limiter.AllowN(t.clock.Now(), 1) && duration < burstSize {
```

Comment:
We shouldn't have a loop here to consume available tokens one by one. This basically means that in order to get better frequency (reducing value of single token and allowing more tokens in rate.Limiter) we will need more loop iterations.

I'm afraid that rate.Limiter doesn't provide any method that allows to take all available tokens in one call.

I think we should drop idea of using rate.Limiter directly, rather try to implement here similar logic to rate.Limiter.advance (https://github.com/golang/time/blob/7e3f01d253248a0a5694eb5b7376dfea18b6397e/rate/rate.go#L364) - it keeps one counter with tokens and timestamp of last token recomputation.

@wojtek-t

Reply:
I initially thought the same, but realized that this is bounded by "burstSize", which is relatively small.

What you're proposing is pretty much what I was proposing in https://github.com/kubernetes/kubernetes/pull/96707
But given it's bounded, I think it may not be that bad...